### PR TITLE
fix: [CI-1382]: update libraries

### PR DESCRIPTION
**Pull request by rafaelmart1ns**
## Describe your changes
bump the golang version to 1.19, redis to v9 and aws-go and add localhost in local config to prevent popup window when ran a server

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49129)

### CODE REVIEW

File: product/log-service/go.mod [hunk 1]
```diff
@@ -1,42 +1,80 @@
 module github.com/harness/harness-core/product/log-service
 
-go 1.14
+go 1.20
```
Comment (lasdox) [inline product/log-service/go.mod:3]: i suggest we set go.mod as 1.19 to not force developers to update to 1.20 right away

Comment (rafaelmart1ns) [reply product/log-service/go.mod:3]: it was changed.

**Comment by rafaelmart1ns** [comment]
trigger gobuild

**Comment by rafaelmart1ns** [comment]
trigger gobuild

**Comment by rafaelmart1ns** [comment]
trigger gobuild

**Comment by rafaelmart1ns** [comment]
trigger gobuild

**Comment by rafaelmart1ns** [comment]
trigger uts

**Comment by rafaelmart1ns** [comment]
trigger uts

File: product/log-service/stream/redis/redis.go [hunk 2]
```diff
@@ -120,28 +119,28 @@ func (r *Redis) Create(ctx context.Context, key string) error {
 	// macro node. MaxLen will always be >= 5000 but can be a few tens of entries
 	// more as well.
 	args := &redis.XAddArgs{
-		Stream:       key,
-		ID:           "*",
-		MaxLenApprox: maxStreamSize,
-		Values:       map[string]interface{}{entryKey: []byte{}},
+		Stream: key,
+		ID:     "*",
+		MaxLen: maxStreamSize,
```
Comment (vistaarjuneja) [inline product/log-service/stream/redis/redis.go:124]: `MaxLenApprox` is more efficient than `MaxLen`. We don't need to use `MaxLen` here.
```However trimming with MAXLEN can be expensive: streams are represented by macro nodes into a radix tree, in order to be very memory efficient. Altering the single macro node, consisting of a few tens of elements, is not optimal. So it's possible to use the command in the following special form:

XADD mystream MAXLEN ~ 1000 * ... entry fields here ...```
From documentation ^

File: product/log-service/stream/redis/redis.go [hunk 3]
```diff
@@ -160,18 +159,18 @@ func (r *Redis) Write(ctx context.Context, key string, lines ...*stream.Line) er
 	for _, line := range lines {
 		bytes, _ := json.Marshal(line)
 		arg := &redis.XAddArgs{
-			Stream:       key,
-			Values:       map[string]interface{}{entryKey: bytes},
-			MaxLenApprox: maxStreamSize,
-			ID:           "*",
+			Stream: key,
+			Values: map[string]interface{}{entryKey: bytes},
+			MaxLen: maxStreamSize,
```
Comment (vistaarjuneja) [inline product/log-service/stream/redis/redis.go:164]: Ditto

File: product/log-service/stream/redis/redis.go [hunk 4]
```diff
@@ -385,7 +384,7 @@ func (r *Redis) setExpiry(key string, expiry time.Duration) error {
 }
 
 // Scan all the keys and set an expiry on them if it's not set
-func (r *Redis) expiryWatcher(expiry time.Duration) {
+func (r *Redis) expiryWatcher(ctx context.Context, expiry time.Duration) {
```
Comment (vistaarjuneja) [inline product/log-service/stream/redis/redis.go:387]: We need to pass ctx to the function which calls `expiryWatcher` as well - can't see that change?

File: product/log-service/config/config.go [hunk 5]
```diff
@@ -25,7 +25,7 @@ type Config struct {
 	}
 
 	Server struct {
-		Bind  string `envconfig:"LOG_SERVICE_HTTP_BIND" default:":8079"`
+		Bind  string `envconfig:"LOG_SERVICE_HTTP_BIND" default:"localhost:8079"`
```
Comment (vistaarjuneja) [inline product/log-service/config/config.go:28]: why is this required and what is the effect?

File: product/log-service/store/s3/s3.go [hunk 6]
```diff
@@ -84,49 +106,52 @@ func (s *Store) Download(ctx context.Context, key string) (io.ReadCloser, error)
...
 		u.PartSize = 32 * 1024 * 1024 // 32MB per part
 	})
 	keyWithPrefix := path.Join("/", s.prefix, key)
-	input := &s3manager.UploadInput{
-		ACL:    aws.String(s.acl),
+	input := &s3.PutObjectInput{
 		Bucket: aws.String(s.bucket),
 		Key:    aws.String(keyWithPrefix),
 		Body:   r,
+		ACL:    types.ObjectCannedACL(s.acl),
```
Comment (vistaarjuneja) [inline product/log-service/store/s3/s3.go:131]: please double check this parameter for `acl` here and why the type has been changed in v2 - and what this means for us (is there anything backward incompatible here). what does `ObjectCannedACL` expect as input

File: product/log-service/store/s3/s3.go [hunk 7]
```diff
@@ -84,49 +106,52 @@ func (s *Store) Download(ctx context.Context, key string) (io.ReadCloser, error)
...
-	req, _ := svc.PutObjectRequest(&s3.PutObjectInput{
+	req, _ := s.presignService.PresignPutObject(ctx, &s3.PutObjectInput{
 		Bucket: aws.String(s.bucket),
 		Key:    aws.String(keyWithPrefix),
+	}, func(opts *s3.PresignOptions) {
+		opts.Expires = time.Duration(int64(expire.Seconds()) * int64(time.Second))
 	})
-	return req.Presign(expire)
+
+	return req.URL, nil
```
Comment (vistaarjuneja) [inline product/log-service/store/s3/s3.go:148]: we should not always return `nil` - return error if it is not nil

File: product/log-service/store/s3/s3.go [hunk 8]
```diff
@@ -9,69 +9,91 @@ package s3
...
+	presignService *s3.PresignClient
 }
 
 // NewEnv returns a new S3 log store from the environment.
 func NewEnv(bucket, prefix, endpoint string, pathStyle bool, accessKeyID, accessSecretKey, region, acl string) *Store {
-	disableSSL := false
-
 	if endpoint != "" {
-		disableSSL = !strings.HasPrefix(endpoint, "https://")
+		http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
```
Comment (vistaarjuneja) [inline product/log-service/store/s3/s3.go:44]: why are we setting `InsecureSkipVerify` as true? This will be susceptible to MITM attacks

File: product/log-service/store/s3/s3.go [hunk 9]
```diff
@@ -84,49 +106,52 @@ func (s *Store) Download(ctx context.Context, key string) (io.ReadCloser, error)
...
+
+	return req.URL, nil
 }
 
 // Delete purges the log stream from the S3 datastore.
 func (s *Store) Delete(ctx context.Context, key string) error {
-	svc := s3.New(s.session)
 	keyWithPrefix := path.Join("/", s.prefix, key)
-	_, err := svc.DeleteObject(&s3.DeleteObjectInput{
+	_, err := s.service.DeleteObject(ctx, &s3.DeleteObjectInput{
```
Comment (vistaarjuneja) [inline product/log-service/store/s3/s3.go:154]: seems like we used to generate a new session for each call but now we are using the same service. can we check that this is concurrency safe with different keys?

File: product/log-service/stream/redis/redis.go (hunk 2 above)
Comment (vistaarjuneja) [reply product/log-service/stream/redis/redis.go:124]: https://redis.io/docs/data-types/streams-tutorial/

File: product/log-service/store/s3/s3.go [hunk 10]
```diff
@@ -84,49 +106,52 @@ func (s *Store) Download(ctx context.Context, key string) (io.ReadCloser, error)
 // DownloadLink creates a pre-signed link that can be used to
 // download the logs to the S3 datastore.
 func (s *Store) DownloadLink(ctx context.Context, key string, expire time.Duration) (string, error) {
-	svc := s3.New(s.session)
 	keyWithPrefix := path.Join("/", s.prefix, key)
-	req, _ := svc.GetObjectRequest(&s3.GetObjectInput{
+	req, _ := s.presignService.PresignGetObject(ctx, &s3.GetObjectInput{
```
Comment (lasdox) [inline product/log-service/store/s3/s3.go:110]: instead of hidding the error ,lets return it if err != nil

File: product/log-service/stream/redis/redis.go (hunk 4 above)
Comment (lasdox) [reply product/log-service/stream/redis/redis.go:387]: yes, @rafaelmart1ns need to change on the scheduler method line 105

File: product/log-service/store/s3/s3.go (hunk 8 above)
Comment (lasdox) [reply product/log-service/store/s3/s3.go:44]: left over probably

File: product/log-service/store/s3/s3.go (hunk 6 above)
Comment (lasdox) [reply product/log-service/store/s3/s3.go:131]: No backward changes, I dont think we should use the constructor here, AWS has predefined the values which we can use, so we should parse and use the constants here.
One thing I noticed is that log-service right now is not using this property, so we should be good for now

File: product/log-service/store/s3/s3.go (hunk 9 above)
Comment (lasdox) [reply product/log-service/store/s3/s3.go:154]: This is not true @vistaarjuneja, old code just created one session for all requests
Sessions have nothing to do with Concurrency 
They are used for scoped configs, since we are not changing configuration, e.g: region, cluster
we should only use one Configuration (similar to session)

File: product/log-service/config/config.go (hunk 5 above)
Comment (rafaelmart1ns) [reply product/log-service/config/config.go:28]: it just prevents the popup window to gave on macos system.

File: product/log-service/store/s3/s3.go (hunk 8 above)
Comment (rafaelmart1ns) [reply product/log-service/store/s3/s3.go:44]: It was updated.

File: product/log-service/store/s3/s3.go (hunk 10 above)
Comment (rafaelmart1ns) [reply product/log-service/store/s3/s3.go:110]: it was updated.

File: product/log-service/store/s3/s3.go (hunk 7 above)
Comment (rafaelmart1ns) [reply product/log-service/store/s3/s3.go:148]: it was updated.

File: product/log-service/stream/redis/redis.go (hunk 2 above)
Comment (rafaelmart1ns) [reply product/log-service/stream/redis/redis.go:124]: Yes, you are right. It was solved by adding the `approx` as true.

File: product/log-service/stream/redis/redis.go (hunk 3 above)
Comment (rafaelmart1ns) [reply product/log-service/stream/redis/redis.go:164]: ditto

File: product/log-service/stream/redis/redis.go (hunk 4 above)
Comment (vistaarjuneja) [reply product/log-service/stream/redis/redis.go:389]: @rafaelmart1ns don't see that change yet? this is invoked on L105. Can you confirm after that change that this thread is running?

File: product/log-service/store/s3/s3.go (hunk 8 above)
Comment (vistaarjuneja) [reply product/log-service/store/s3/s3.go:45]: Don't see the updated change

File: product/log-service/store/s3/s3.go [hunk 11]
```diff
@@ -84,49 +107,58 @@ func (s *Store) Download(ctx context.Context, key string) (io.ReadCloser, error)
...
 // upload the logs to the S3 datastore.
 func (s *Store) UploadLink(ctx context.Context, key string, expire time.Duration) (string, error) {
-	svc := s3.New(s.session)
 	keyWithPrefix := path.Join("/", s.prefix, key)
-	req, _ := svc.PutObjectRequest(&s3.PutObjectInput{
+	req, err := s.presignService.PresignPutObject(ctx, &s3.PutObjectInput{
 		Bucket: aws.String(s.bucket),
 		Key:    aws.String(keyWithPrefix),
+	}, func(opts *s3.PresignOptions) {
+		opts.Expires = time.Duration(int64(expire.Seconds()) * int64(time.Second))
```
Comment (vistaarjuneja) [inline product/log-service/store/s3/s3.go:149]: `expire` is of `time.Duration` so you don't need this conversion

File: product/log-service/store/s3/s3.go (hunk 6 above)
Comment (vistaarjuneja) [reply product/log-service/store/s3/s3.go:135]: @rafaelmart1ns can you confirm on the above once? Do double check whether we have acl set in `ng-prod-manifests` anywhere and whether we need the constructor

File: product/log-service/store/s3/s3.go (hunk 9 above)
Comment (vistaarjuneja) [reply product/log-service/store/s3/s3.go:161]: @lasdox you are right - we created one session but looks like service was created each time. But now we use a single service. 
This is what I was looking for: https://aws.github.io/aws-sdk-go-v2/docs/making-requests/#concurrently-using-service-clients and looks like this is ok. Can you guys also go through this once cc: @rafaelmart1ns
